ethereumj-core/src/main/java/org/ethereum/sync/FastSyncManager.java

```diff
@@ -191,27 +184,19 @@ public TrieNodeRequests(TrieNodeRequest request) {
     }
 
     Deque<TrieNodeRequest> nodesQueue = new LinkedBlockingDeque<>();
-    ByteArrayMap<TrieNodeRequests> pendingNodes = new ByteArrayMap<>();
+    ByteArrayMap<TrieNodeRequest> pendingNodes = new ByteArrayMap<>();
     ByteArrayMap<Long> knownHashes = new ByteArrayMap<>();
     Long requestId = 0L;
 
     private synchronized void purgePending(byte[] hash) {
-        boolean needToRemove = true;
-        TrieNodeRequests requests = pendingNodes.get(hash);
-        if (requests == null) return;
-        for (TrieNodeRequest request: requests.requests) {
-            if (!request.requestSent.isEmpty()) {
-                needToRemove = false;
-                break;
-            }
-        }
-        if (needToRemove) pendingNodes.remove(hash);
+        TrieNodeRequest request = pendingNodes.get(hash);
+        if (request.requestSent.isEmpty()) pendingNodes.remove(hash);
     }
 
     synchronized void processTimeouts() {
         long cur = System.currentTimeMillis();
-        for (TrieNodeRequests nodeRequests : new ArrayList<>(pendingNodes.values())) {
-            for (TrieNodeRequest request: nodeRequests.requests) {
+        for (TrieNodeRequest request : new ArrayList<>(pendingNodes.values())) {
+            synchronized (request.requestSent) {
```

Comment:
That's why `ConcurrentHashMap` made it work (it internally synchronizes on its instance)